_source/_posts/2020-04-20-android-authentication.adoc

```diff
@@ -0,0 +1,508 @@
+private var keyguardEncryptionManager: GuardedEncryptionManager? = null
+
+override fun onResume() {
+   super.onResume()
+
+   // Check whether the user has already authenticated using the device's authentication method. If it hasn't, ask them to do it
+   if (currentEncryptionManager?.isUserAuthenticatedOnDevice?.not() == true) {
+       showKeyguard()
+   }
+}
+
+private fun showKeyguard() {
+   Biometric(
+       fragmentActivity = this,
+       onSuccessListener = {
+           Log.d("MainActivity", "Biometric authentication succeeded")
+           Toast.makeText(this, "Biometric authentication succeeded", Toast.LENGTH_SHORT).show()
+       },
+       onCancelListener = {
+           runOnUiThread {
+               Log.d("MainActivity", "Biometric authentication cancelled")
+               Toast.makeText(this, "Biometric authentication cancelled", Toast.LENGTH_SHORT).show()
+               finish()
+           }
+       },
+       onErrorListener = { code, message ->
+           runOnUiThread {
+               Log.d("MainActivity", "Biometric authentication failed")
+               Toast.makeText(this, "Biometric authentication failed", Toast.LENGTH_SHORT).show()
+               finish()
+           }
+       }).show()
+}
+
+private fun setupWebAuth() {
+   keyguardEncryptionManager = GuardedEncryptionManager(this, Int.MAX_VALUE)
+
+   webAuth = WebAuthBuilder()
+       .withConfig(oidcConfig)
+       .withContext(applicationContext)
+       .withCallbackExecutor(null)
+       .withEncryptionManager(keyguardEncryptionManager)
+       .create()
+   sessionClient = webAuth.sessionClient
+}
+----
+
+Now, whenever a user opens a new instance of your app, they will be able to access it using their biometric information. In the below example, a Google Pixel phone, the authentication is a  fingerprint:
+
+image::{% asset_path 'blog/android-authentication/android-fingerprint-prompt.png' %}[alt=Android fingerprint prompt,width=400,align=center]
+
+== Summary
+
+It's now easier than ever to implement OAuth 2.0 and OpenID authorization thanks to Okta OIDC SDK. A high effort, high maintenance chore just became a very straightforward task.
+
+By taking advantage of `BiometricPrompt` and Kotlin, we can make our apps even safer, while still creating a seamless user experience that  feels native to each manufacturer's UI style.
+
+Although the example we've created here does enough to satisfy the needs of the majority of apps, Okta OIDC SDK doesn't stop there. The https://github.com/okta/okta-oidc-android[Okta OIDC Android repository] contains  a variety of ideas and suggestions to improve user experience like:
+
+- Using your own OkHttp client
+- Using a custom UI to log in
+- Having fine-grained control over session tokens' expiration, refresh, etc...
+- Settings to handle preference of browser client for the authentication process (Chrome, Samsung browser, Firefox, etc...)
+
+== Further learning
```

Comment:
This is pretty generic. I'd change it to be something like "Learn More About Android Auth and OIDC"